### PR TITLE
Reject process instance migration for called process instance

### LINKED ISSUE #15126: Reject process instance migration for called process instance
In the `ProcessInstanceMigration` command processor, reject the command if:
- the migrated process instance is a child process instance, i.e. it is created by a call activity in a parent process instance

Rejecting a command includes both appending a rejection with the rejection writer, as well as responding the rejection with the response writer.

Some tests must showcase this behavior.

Blocked by:
- https://github.com/camunda/zeebe/issues/15111

### PR DESCRIPTION
**Pull request by berkaycanbc**
## Related issues

closes #15126 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationMigrateProcessor.java [hunk 1]
```diff
@@ -55,6 +55,7 @@ public class ProcessInstanceMigrationMigrateProcessor
   private static final String ERROR_MESSAGE_PROCESS_DEFINITION_NOT_FOUND =
       "Expected to migrate process instance to process definition but no process definition found with key '%d'";
 
+  private static final long NO_PARENT = -1L;
```
Comment (koevskinikola) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationMigrateProcessor.java:58]: :+1: I like how you define these constants to improve the readability of the code.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationMigrateProcessor.java [hunk 2]
```diff
@@ -100,6 +101,10 @@ public void processRecord(final TypedRecord<ProcessInstanceMigrationRecord> comm
       return;
     }
 
+    if (processInstance.getValue().getParentProcessInstanceKey() != NO_PARENT) {
+      throw new ChildProcessMigrationException(processInstanceKey);
+    }
```
Comment (koevskinikola) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationMigrateProcessor.java:106]: :+1: Pretty nice that you perform this check early.

**Review by koevskinikola — APPROVED** [review]
LGTM @berkaycanbc! :rocket: 

:wrench: It would be good if we can extract all of these rejection checks in some sort of "checker" or "validator" class, but this is not in the scope of this issue.

**Comment by berkaycanbc** [comment]
> LGTM @berkaycanbc! 🚀
> 
> 🔧 It would be good if we can extract all of these rejection checks in some sort of "checker" or "validator" class, but this is not in the scope of this issue.

I definitely agree. It just grown too much. We should refactor these after the release. 👍

**Comment by berkaycanbc** [comment]
bors merge

**Pull request merged**